### DISCUSSION TITLE
BMP280 sensors?

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#12405] [Hardware & Peripherals] [answered]
Note that i will be using a raspberry Pi PICO for this

The last D18B20 senors i got seem to have been bad clones (only give CRC errors) so i figure i can try a different sensor and found these

* https://www.ebay.com/itm/152712156737

i have gotten stuff from this seller before and the only issue i had was the D18B20 sensors, everything else has been good and fake D18B20s are everywhere and the seller did deal with the situation in a timely manner

* I assume there is a good/recommended library to use these? link? (both SPI and I2C)
* How long can the wires be? (could use ethernet or telephone wire)

**Comment by robert-hh** [comment, marked as answer]
I adapted am early Adafruit driver: https://github.com/robert-hh/BME280. It supports both I2C and SPI.
The BME280 works well. In a test of various sensors, the temperature and pressure values were always be similar within the given error range. Only the humidity values varied a lot. These difference got smaller after keeping the sensors at the same place for several months.
The wire lengths for I2C and SPI are meant to be short, less than 30cm. You can extend that by lowering the clock rate and for I2C by making the pull-up resistor as small as possible, down to about 1k Ohm. The data sheet does not tell how much current a BME280 can source or sink. At 1k it would be ~3mA. You may use lower values. Using shielded wires can help. Telephone or Internet wires are good enough. You have to try the length. I recall that people in the forum used like 5m long wires. Generally it should be less critical than a DS18B20, since the baud rate and impedance can be lowered. At DS18B20, the rate is fixed.

**Reply by robert-hh** [reply]
Note: The BME280 and BMP280 are compatible with respect to the register and communication interface. Only that with the BMP280 the reading of humidity is not supported. It may return no values or nonsense values.

**Reply by robert-hh** [reply]
I just tried the BME280 driver with a BMP280 breakout. It works and returns proper values for  temperature and pressure. The humidity value is returned as 0.

**Reply by GM-Script-Writer-62850** [reply]
is there a way to get a humidity reading? or is the BMP280 IC only offer temp/pressure? and BMP280 IC offer temp/humidity/presure

**Reply by robert-hh** [reply]
The BMP280 returns only temperature and pressure. For additional humidity values you have to use the BME280.

**Reply by GM-Script-Writer-62850** [reply]
i just got a few BME280s, very surprised with now much lower the humidity readings are using this compared to my DHT22s (well over 10%)

i have I2C working good, but how do i connect it for SPI?, i have 6 pins  VCC, GND, SCL, SDA, CSB, and SDO, and the PICO has SCK, TX, RX, AND CSn, note note how to connect the data wires

**Reply by robert-hh** [reply]
CSB is Chip select to Pico CS
SCL is SPI Clock to Pico SCK
SDA is MISO to Pico TX
SDO is MOSI to Pico RX

If that fails, swap TX and RX. The definitions of the Pin are not always clear about whether it's the device or host view.
Humidity readings are always somewhat erratic. BME280's get better after a while. I made a series of tests, trying to get a well defined humidity environment using saturated air over a few chemicals.

![test results](https://github.com/micropython/micropython/assets/12476868/83b8540c-f6de-4825-af60-d5c85a012fbf)

**Reply by GM-Script-Writer-62850** [reply]
Thanks, though you used `Pico SCK` twice, but CSB is rather obvious
Looks like the library you linked is I2C only...

**Reply by robert-hh** [reply]
Yes, it's I2C only. The only advantage in using SPI would be the actively driven signal lines. But it's two more wires.

**Reply by GM-Script-Writer-62850** [reply]
I may end up also using 8 ADS1115 boards, that would use 100% of my I2C capacity

also want to try and reproduce a issue i found using my W5500 NIC to see if it is a W5500 issue or a SPI issue

**Reply by GM-Script-Writer-62850** [reply]
i found this lib: https://pypi.org/project/bme280-upy/

The I2C gives the same data as the lib you linked, however when using SPI i different results

```
I2C: 23.3 *C; 52.5 % rel. hum.; 1019.3 hPa
SPI: 22.9 *C; 90.5 % rel. hum.; 658.8 hPa
```

My DHT22s nearest me are reading 65.7%, 62.4%,and 68.1%

EDIT:

SCL -> SCK (GP2)
SDA -> TX (GP3)
SDO -> RX (GP4)
SCB -> CSn (GP5)

**Reply by robert-hh** [reply]
The calculation of the humidity in that implementation differs from formula given in the data sheet. The calculation for the pressure seems right, but the value is off as well. Besides that, it's an overcomplicated piece of code.

**Reply by GM-Script-Writer-62850** [reply]
the issue seems the 1st  pass can give bad data, all subsequent runs give good data

Note these are no more than 5cm apart with the D18B20 in the middle (2 second delay, not i used the read time of the dht22 as a delay after using convert temp for the d18b20)
```
BME280: 22.8 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.75 *C

BME280: 22.7 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.8125 *C

BME280: 22.8 *C; 52.3 %; 1021.2 hPa
DHT22:  22.4 *C; 62.6 %
D18B20: 22.8125 *C

BME280: 22.8 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.75 *C

BME280: 22.7 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.75 *C

BME280: 22.7 *C; 52.3 %; 1021.2 hPa
DHT22:  22.4 *C; 62.6 %
D18B20: 22.75 *C

BME280: 22.7 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.75 *C

BME280: 22.7 *C; 52.2 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.8125 *C

BME280: 22.7 *C; 52.2 %; 1021.2 hPa
DHT22:  22.4 *C; 62.6 %
D18B20: 22.8125 *C

BME280: 22.7 *C; 52.3 %; 1021.2 hPa
DHT22:  22.5 *C; 62.6 %
D18B20: 22.75 *C
```

**Reply by robert-hh** [reply]
Similar findings than I had. The temperature is usually in range, give that the precision is specified as +/- 0.5°C. The humidity of the BME280 seemd always a little bit off, even with a +/- 2% error range. The air pressure of the BME280 was always on spot, compared to the reading of a close-by University site at about the same height.
